#7997: Migrated Input and Delete features to beforeinput

oleq opened the pull request:
### Suggested merge commit message (convention)

Feature: Migrated `Input`, `Delete`, and `Enter` features to the `beforeinput` event. Closes #7997. Closes #7998. Closes #7999.

BREAKING CHANGE (engine): The [`MutationObserver`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_observer_mutationobserver-MutationObserver.html) may no longer be [loaded by the editing view](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/architecture/editing-engine.html#observers). In web browsers that support DOM Input Events, its responsibilities have been taken over by [`InputObserver`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_observer_inputobserver-InputObserver.html).

BREAKING CHANGE (engine): The [`mutations`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:mutations) event is no longer fired by the view document in web browsers that support DOM Input Events. To react to low-level changes in the document, use the [`beforeinput`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:beforeinput) event instead, or better yet, take advantage of [`delete`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:delete) and [`insertText`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:insertText) events.

BREAKING CHANGE (typing): The [InputCommand](https://ckeditor.com/docs/ckeditor5/latest/api/module_typing_inputcommand-InputCommand.html) has been deprecated. Use the [`InsertTextCommand`](https://ckeditor.com/docs/ckeditor5/latest/api/module_typing_inserttextcommand-InsertTextCommand.html) instead (`editor.execute( 'insertText', options )`).

BREAKING CHANGE (typing): The [`@ckeditor/ckeditor5-typing/src/utils/injectunsafekeystrokeshandling`](https://ckeditor.com/docs/ckeditor5/latest/api/module_typing_utils_injectunsafekeystrokeshandling.html) module containing typing helpers (e.g. [`isNonTypingKeystroke()`](https://ckeditor.com/docs/ckeditor5/latest/api/module_typing_utils_injectunsafekeystrokeshandling.html#function-isNonTypingKeystroke)) is now available under `@ckeditor/ckeditor5-typing/src/utils/input/utils`. Please update import paths in your projects to avoid import errors.

niegowski reviewed (commented):
Partial review

oleq commented on packages/ckeditor5-enter/src/enterobserver.js:
```diff
@@ -20,27 +21,89 @@ export default class EnterObserver extends Observer {
 			}
+
+			this._fireEnterEvent( domEvent, evt.stop, isSoftEnter );
+			data.preventDefault();
```
We discussed this F2F and this is to stay on the safe side and leave no room for the browser to act.

oleq commented on packages/ckeditor5-typing/src/inserttextcommand.js:
```diff
@@ -0,0 +1,117 @@
+			model.deleteContent( selection );
+
+			if ( text ) {
+				const modelText = writer.createText( text, modelDocument.selection.getAttributes() );
```
It was a copy-paste from `InputCommand` and I didn't change its implementation as it's beyond the scope of this PR. Maybe we can have a follow-up for this?

niegowski replied:
ok

oleq commented on packages/ckeditor5-typing/src/utils/delete/injectbeforeinputdeletehandling.js:
```diff
@@ -0,0 +1,104 @@
+	viewDocument.on( 'delete', ( evt, data ) => {
+		const { direction, sequence, selectionToRemove, inputType, unit } = data;
+
+		// Both "deleteContentBackward" and "deleteContentForward" must operate on the unit-level despite the
+		// editing view range available and, in case of other input types, used efficiently to delete content.
+		// This is related to the multi-byte characters decomposition (like complex emojis). Check out
+		// the comments in DeleteObserver's DELETE_EVENT_TYPES to learn more.
+		if ( inputType === 'deleteContentBackward' ) {
+			const deleteCommandData = {
+				sequence,
+				unit
+			};
+
+			// Android IMEs have a quirk which is addressed by passing a complete selection in case of
+			// the 'deleteContentBackward' event type. See DeleteObserver#_enableBeforeInputBasedObserver()
+			// to learn more.
+			if ( env.isAndroid ) {
+				deleteCommandData.selection = selectionToRemove;
+			}
+
+			editor.execute( 'delete', deleteCommandData );
+		} else if ( inputType === 'deleteContentForward' ) {
+			editor.execute( 'forwardDelete', {
+				unit,
+				sequence
+			} );
+		}
```
The condition would also kick in on `deleteWordForward` and `deleteWordBackward`.

niegowski replied:
So maybe based on `unit == 'character' || unit == 'codePoint'` ? Checking `inputType` this level seems like a violation of abstraction. Also `delete` next to `inputType` looks strange (it's internal that it comes from `beforeInput`)

oleq commented on packages/ckeditor5-typing/src/utils/delete/injectbeforeinputdeletehandling.js:
```diff
@@ -0,0 +1,104 @@
+		editingView.scrollToTheSelection();
+	} );
+
+	// Android IMEs have a quirk - they change DOM selection after the input changes were performed by the browser.
+	// This happens on `keyup` event. Android doesn't know anything about our deletion and selection handling.
+	// Even if the selection was changed during input events, IME remembers the position where the selection "should"
+	// be placed and moves it there.
+	//
+	// To prevent incorrect selection, we save the selection after deleting here and then re-set it on `keyup`.
+	// This has to be done on DOM selection level, because on `keyup` the model selection is still the same as it was
+	// just after deletion, so it wouldn't be changed and the fix would do nothing.
+	//
+	// **Note**: See DeleteObserver#_enableBeforeInputBasedObserver() for the first part of this quirk.
+	if ( env.isAndroid ) {
```
I don't like this code either but this would result in `DeleteObserver` listening to its own event (`#delete`), which is against the single-responsibility principle. And it would look weird 😛 TBH I'm not even 100% sure it works at this moment. I'm not saying no, but let's get back to this code in #8011 maybe?

niegowski replied:
Ok.

niegowski commented on packages/ckeditor5-typing/src/utils/delete/injectbeforeinputdeletehandling.js:
```diff
@@ -21,46 +21,28 @@ export default function injectBeforeInputDeleteHandling( editor ) {
+		//   to learn more.
+		// * Android IMEs have a quirk which is addressed by passing a complete selection in case of
+		//   the "codePoint" unit. See DeleteObserver#_enableBeforeInputBasedObserver() to learn more.
+		if ( ( unit !== 'codePoint' && unit !== 'character' ) || ( env.isAndroid && unit === 'codePoint' ) ) {
```
```suggestion
		if ( unit !== 'codePoint' && unit !== 'character' || env.isAndroid && unit === 'codePoint' ) {
```

oleq commented:
Feature (typing): Migrated `Input` and `Delete` features to the `beforeinput` event. Implemented `InsertTextCommand` and the `insertText` view document event. Closes [#7997](https://github.com/ckeditor/ckeditor5/issues/7997). Closes [#7998](https://github.com/ckeditor/ckeditor5/issues/7998). 

Feature (enter): Migrated the `Enter` feature (and `EnterObserver`) to the `beforeinput` event. Closes [#7999](https://github.com/ckeditor/ckeditor5/issues/7999).

Other (engine): `MutationObserver` should only be loaded when the web browser does not support Input Events. Closes [#7999](https://github.com/ckeditor/ckeditor5/issues/7999).

Other (engine): Added the `isComposing` property to the `beforeinput` event data (see [#7997](https://github.com/ckeditor/ckeditor5/issues/7997)).

Tests (block-quote): Aligned feature tests to `beforeinput`\-driven delete (see [#7998](https://github.com/ckeditor/ckeditor5/issues/7998)).

Internal (engine): Aligned the `SelectionObserver` to the `beforeinput`\-driven typing (see [#799](https://github.com/ckeditor/ckeditor5/issues/7998)7).

Internal (link): Aligned the `LinkEditing` feature to the `beforeinput`\-driven deleting. Used the `InsertTextCommand` instead of `InputCommand` in the `AutoLink` plugin. (see [#799](https://github.com/ckeditor/ckeditor5/issues/7998)7, [#7998](https://github.com/ckeditor/ckeditor5/issues/7998)).

Tests (list): Used `InsertTextCommand` instead of deprecated `InputCommand` (see [#799](https://github.com/ckeditor/ckeditor5/issues/7998)7).

Tests (media-embed): Used `InsertTextCommand` instead of deprecated `InputCommand` (see [#799](https://github.com/ckeditor/ckeditor5/issues/7998)7).

Internal (restricted-editing): Used `InsertTextCommand` instead of deprecated `InputCommand` (see [#799](https://github.com/ckeditor/ckeditor5/issues/7998)7).

Internal (special-characters): Used `InsertTextCommand` instead of deprecated `InputCommand` (see [#799](https://github.com/ckeditor/ckeditor5/issues/7998)7).

Internal (table): Aligned the `TableSelection` plugin to the `beforeinput`\-driven typing (see [#799](https://github.com/ckeditor/ckeditor5/issues/7998)7).

Internal (widget): Aligned the `WidgetTypeAround` plugin to the `beforeinput`\-driven typing (see [#799](https://github.com/ckeditor/ckeditor5/issues/7998)7).

BREAKING CHANGE (engine): The [`MutationObserver`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_observer_mutationobserver-MutationObserver.html) may no longer be [loaded by the editing view](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/architecture/editing-engine.html#observers). In web browsers that support DOM Input Events, its responsibilities have been taken over by [`InputObserver`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_observer_inputobserver-InputObserver.html).

BREAKING CHANGE (engine): The [`mutations`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:mutations) event is no longer fired by the view document in web browsers that support DOM Input Events. To react to low-level changes in the document, use the [`beforeinput`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:beforeinput) event instead, or better yet, take advantage of [`delete`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:delete) and [`insertText`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_document-Document.html#event-event:insertText) events.

BREAKING CHANGE (typing): The [InputCommand](https://ckeditor.com/docs/ckeditor5/latest/api/module_typing_inputcommand-InputCommand.html) has been deprecated. Use the [`InsertTextCommand`](https://ckeditor.com/docs/ckeditor5/latest/api/module_typing_inserttextcommand-InsertTextCommand.html) instead (`editor.execute( 'insertText', options )`).

BREAKING CHANGE (typing): The [`@ckeditor/ckeditor5-typing/src/utils/injectunsafekeystrokeshandling`](https://ckeditor.com/docs/ckeditor5/latest/api/module_typing_utils_injectunsafekeystrokeshandling.html) module containing typing helpers (e.g. [`isNonTypingKeystroke()`](https://ckeditor.com/docs/ckeditor5/latest/api/module_typing_utils_injectunsafekeystrokeshandling.html#function-isNonTypingKeystroke)) is now available under `@ckeditor/ckeditor5-typing/src/utils/input/utils`. Please update import paths in your projects to avoid import errors.
